File: apps/runtime/src/api/tests/language.rs
//! `GET …/language-service` and the session routes over HTTP.
//!
//! The probe is deliberately written so it never depends on what this machine
//! has installed: what is asserted is that every language is listed, that
//! "not usable" always carries a reason, and that the execute gate is applied
//! to the answer rather than to whether the answer exists.

use axum::http::StatusCode;
use serde_json::{Value, json};

use super::support::{call, router_fixture};

async fn workspace(router: &axum::Router, root: &std::path::Path) -> String {
    let (status, workspace) = call(
        router,
        "POST",
        "/api/workspaces",
        Some(json!({ "name": "语言", "rootPath": root.to_string_lossy() })),
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    workspace["id"].as_str().unwrap().to_owned()
}

fn rows(probe: &Value) -> Vec<&Value> {
    probe["servers"].as_array().unwrap().iter().collect()
}

#[tokio::test]
async fn every_language_is_listed_and_anything_unusable_says_why() {
    let (router, directory) = router_fixture("language-probe").await;
    let id = workspace(&router, directory.path()).await;

    let (status, probe) = call(
        &router,
        "GET",
        &format!("/api/workspaces/{id}/language-service"),
        None,
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    // One row per language in the registry — a missing server is listed as
    // missing, because "we looked and it is not there" is what the settings
    // page has to be able to say.
    assert_eq!(
        rows(&probe).len(),
        crate::language::registry::languages().len()
    );
    for row in rows(&probe) {
        assert!(!row["serverId"].as_str().unwrap().is_empty());
        assert!(!row["languageId"].as_str().unwrap().is_empty());
        assert!(!row["fileExtensions"].as_array().unwrap().is_empty());
        if row["state"] == "unsupported" {
            let reason = row["reason"].as_str().unwrap_or_default();
            assert!(
                [
                    "server_not_found",
                    "server_probe_failed",
                    "execution_not_granted",
                    "disabled",
                ]
                .contains(&reason),
                "unexpected reason {reason}"
            );
        }
        // A server that is not running has no pid at all. A pid of 0 would be
        // a process the resource panel could try to claim.
        if row["state"] != "running" {
            assert!(row["pid"].is_null(), "{row}");
        }
    }
}

#[tokio::test]
async fn a_workspace_without_execute_may_still_look_but_starts_nothing() {
    let (router, directory) = router_fixture("language-execute").await;
    let id = workspace(&router, directory.path()).await;
    let (status, _) = call(
        &router,
        "PATCH",
        &format!("/api/workspaces/{id}"),
        Some(json!({ "permissions": { "read": true, "write": true, "execute": false } })),
    )
    .await;
    assert_eq!(status, StatusCode::OK);

    let (status, probe) = call(
        &router,
        "GET",
        &format!("/api/workspaces/{id}/language-service"),
        None,
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    // The probe itself is the runtime's own `--version`, not the project's
    // code, so it still runs. What changes is the answer.
    assert_eq!(probe["status"], "unavailable");
    assert_eq!(probe["reason"], "execution_not_granted");
    for row in rows(&probe) {
        assert_eq!(row["state"], "unsupported", "{row}");
        assert!(row["pid"].is_null());
    }

    // And opening a session is refused before anything is spawned.
    let (status, error) = call(
        &router,
        "POST",
        &format!("/api/workspaces/{id}/language/sessions"),
        Some(json!({ "languageId": "python", "clientId": "node-1" })),
    )
    .await;
    assert_eq!(status, StatusCode::FORBIDDEN);
    assert_eq!(error["code"], "forbidden");
}

#[tokio::test]
async fn a_language_nobody_serves_is_refused_by_name() {
    let (router, directory) = router_fixture("language-unknown").await;
    let id = workspace(&router, directory.path()).await;
    // A workspace is not executable by default; the point of this test is the
    // language name, so the grant is given first.
    let (status, _) = call(
        &router,
        "PATCH",
        &format!("/api/workspaces/{id}"),
        Some(json!({ "permissions": { "read": true, "write": true, "execute": true } })),
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    let (status, error) = call(
        &router,
        "POST",
        &format!("/api/workspaces/{id}/language/sessions"),
        Some(json!({ "languageId": "cobol", "clientId": "node-1" })),
    )
    .await;
    assert_eq!(status, StatusCode::BAD_REQUEST);
    assert_eq!(error["message"], "language_unknown");
}

#[tokio::test]
async fn a_session_id_nobody_owns_is_not_a_capability() {
    let (router, directory) = router_fixture("language-session-id").await;
    let id = workspace(&router, directory.path()).await;
    for (method, path) in [
        (
            "DELETE",
            format!("/api/workspaces/{id}/language/sessions/made-up"),
        ),
        (
            "POST",
            format!("/api/workspaces/{id}/language/servers/made-up/stop"),
        ),
    ] {
        let (status, _) = call(&router, method, &path, None).await;
        assert!(
            status == StatusCode::OK || status == StatusCode::NOT_FOUND,
            "{path} answered {status}"
        );
    }
    // Applying an edit through a session that does not exist finds nothing to
    // apply it with, rather than falling back to applying it anyway.
    let (status, _) = call(
        &router,
        "POST",
        &format!("/api/workspaces/{id}/language/sessions/made-up/edits"),
        Some(json!({ "edit": { "changes": {} }, "expectedSha256": {} })),
    )
    .await;
    assert_eq!(status, StatusCode::NOT_FOUND);
}

#[tokio::test]
async fn a_read_only_workspace_cannot_apply_a_language_edit() {
    let (router, directory) = router_fixture("language-edit-gate").await;
    let id = workspace(&router, directory.path()).await;
    let (status, _) = call(
        &router,
        "PATCH",
        &format!("/api/workspaces/{id}"),
        Some(json!({ "permissions": { "read": true, "write": false, "execute": true } })),
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    let (status, error) = call(
        &router,
        "POST",
        &format!("/api/workspaces/{id}/language/sessions/any/edits"),
        Some(json!({ "edit": { "changes": {} }, "expectedSha256": {} })),
    )
    .await;
    // The write gate is checked before the session is looked up: a read-only
    // workspace is refused whether or not the session exists.
    assert_eq!(status, StatusCode::FORBIDDEN);
    assert_eq!(error["code"], "forbidden");
}

File: packages/shared/src/api/language.ts
import { z } from "zod";

/**
 * Editor language services — the Runtime ↔ Web half of
 * `docs/design/language-service.md` §2.9.
 *
 * The shapes mirror `proto/armadra/v1/language.proto` field for field, because
 * a remote workspace answers the same questions over the Worker protocol and
 * the browser must not be able to tell which machine replied.
 *
 * Two rules the schemas encode rather than document:
 *
 *  * **`unsupported` is an answer, not an absence.** Every descriptor that is
 *    not `available` carries a stable `reason` key, so the settings page can
 *    say *what* is missing instead of showing an empty row.
 *  * **The browser never sees an absolute path.** Every uri inside
 *    `serverCapabilities` and inside a `WorkspaceEdit` is `armadra:///<rel>`;
 *    `executable` is the one absolute path here, and it names a program on the
 *    execution host, not a file in the project.
 */

/** Where one server is in its life (design §1.3). */
export const languageServerStateSchema = z.enum([
  "available",
  "unsupported",
  "starting",
  "running",
  "idleStopped",
  "crashed",
  "stopped",
  "disconnected",
]);

/**
 * Why a server is not usable. Stable keys the interface localises; a runtime
 * that grows a new one is rendered as generic text rather than crashing the
 * settings page, which is why this is a string and not an enum.
 */
export const LANGUAGE_UNSUPPORTED_REASONS = [
  "server_not_found",
  "server_probe_failed",
  "execution_not_granted",
  "disabled",
  "language_unknown",
  "too_many_servers",
  "containment_unavailable",
  "resource_exhausted",
  "unsupported_remote",
] as const;

/** What a running server actually answers, from its own capabilities. */
export const languageFeatureSchema = z.enum([
  "completion",
  "diagnostics",
  "hover",
  "definition",
  "references",
  "rename",
  "formatting",
  "documentSymbol",
  "workspaceSymbol",
  "codeAction",
  "signatureHelp",
]);

export const languageServerDescriptorSchema = z.object({
  serverId: z.string(),
  languageId: z.string(),
  fileExtensions: z.array(z.string()),
  /** Absolute path on the execution host; empty when nothing was found. */
  executable: z.string(),
  version: z.string(),
  state: languageServerStateSchema,
  reason: z.string().optional(),
  features: z.array(languageFeatureSchema),
  restartCount: z.number().int().nonnegative(),
  /** Only while the process is running: pid + start time is the identity. */
  pid: z.number().int().nullish(),
  startTimeUnixMs: z.number().int().nullish(),
  openDocuments: z.number().int().nonnegative(),
  probedAtUnixMs: z.number().int().nonnegative(),
});

/**
 * `GET /api/workspaces/{id}/language-service` (`?refresh=1` re-probes).
 *
 * `unavailable` no longer means "Armadra has no LSP": it means this workspace
 * cannot run one right now, and `reason` says why. The per-language rows are
 * listed either way, so the settings page can show what is missing rather than
 * an empty panel.
 */
export const languageServiceStatusSchema = z.object({
  status: z.enum(["available", "unavailable"]),
  reason: z.string().optional(),
  executionHostId: z.string().default("local"),
  servers: z.array(languageServerDescriptorSchema).default([]),
});

export const openLanguageSessionRequestSchema = z.object({
  languageId: z.string().min(1).max(60),
  clientId: z.string().min(1).max(120),
  /** The client's LSP `ClientCapabilities`, intersected on the host. */
  clientCapabilities: z.unknown().optional(),
});

export const openLanguageSessionResponseSchema = z.object({
  sessionId: z.string(),
  generation: z.number().int().nonnegative(),
  serverId: z.string(),
  state: languageServerStateSchema,
  reason: z.string().optional(),
  /** The server's own `InitializeResult.capabilities`, replayed verbatim. */
  serverCapabilities: z.unknown().optional(),
});

/**
 * `POST …/language/sessions/{sessionId}/edits`.
 *
 * A path that is absent from `expectedSha256` must not already exist. An empty
 * string would be a version, and versions are checked; absence is the way to
 * say "create only".
 */
export const applyLanguageEditRequestSchema = z.object({
  edit: z.unknown(),
  expectedSha256: z.record(z.string(), z.string().regex(/^[a-f0-9]{64}$/)),
});

export const languageAppliedFileSchema = z.object({
  path: z.string(),
  sha256: z.string().regex(/^[a-f0-9]{64}$/),
  size: z.number().int().nonnegative(),
});

export const languageFailedFileSchema = z.object({
  path: z.string(),
  code: z.string(),
  message: z.string(),
});

/**
 * Writing stops at the first failure, so a partial application is a real
 * outcome and both lists come back: the dialog says which files changed.
 */
export const applyLanguageEditResultSchema = z.object({
  applied: z.array(languageAppliedFileSchema),
  failed: z.array(languageFailedFileSchema),
});

/** `$/progress` reduced to what a status line can show. */
export const languageProgressSchema = z.object({
  percent: z.number().int().min(0).max(100).nullish(),
  title: z.string(),
});

/** Event `language.session` — the status line and the settings page. */
export const languageSessionEventSchema = z.object({
  workspaceId: z.string(),
  sessionId: z.string(),
  serverId: z.string(),
  generation: z.number().int().nonnegative(),
  state: languageServerStateSchema,
  reason: z.string().optional(),
  restartCount: z.number().int().nonnegative(),
  progress: languageProgressSchema.nullish(),
});

/** Event `language.server` — pushed on probe and on every state change. */
export const languageServerEventSchema = z.object({
  workspaceId: z.string(),
  executionHostId: z.string(),
  server: languageServerDescriptorSchema,
  /**
   * The tail of the server's stderr, redacted, for a crash the settings page
   * has to explain. Never persisted and never logged.
   */
  stderrTail: z.string().optional(),
});

export type LanguageServerState = z.infer<typeof languageServerStateSchema>;
export type LanguageFeature = z.infer<typeof languageFeatureSchema>;
export type LanguageServerDescriptor = z.infer<
  typeof languageServerDescriptorSchema
>;
export type LanguageServiceStatus = z.infer<typeof languageServiceStatusSchema>;
export type OpenLanguageSessionRequest = z.infer<
  typeof openLanguageSessionRequestSchema
>;
export type OpenLanguageSessionResponse = z.infer<
  typeof openLanguageSessionResponseSchema
>;
export type ApplyLanguageEditRequest = z.infer<
  typeof applyLanguageEditRequestSchema
>;
export type ApplyLanguageEditResult = z.infer<
  typeof applyLanguageEditResultSchema
>;
export type LanguageSessionEvent = z.infer<typeof languageSessionEventSchema>;
export type LanguageServerEvent = z.infer<typeof languageServerEventSchema>;
